### DISCUSSION TITLE
Dynamic filter

### DISCUSSION THREAD

**Discussion by mattiason** [Budibase/budibase#5493] [Help]
I want to build a form with filters for relations.

I have another relation to another table which contains a value I want to filter on.

So what I want to is to have a relationship picker select a record and the use another field of that record to filter next relationship picker. 

Example from relationship picker 1 table.a.fieldB should filter to only show records in relationship picker 2  that are table.a.fieldB= table.b.fieldC

I hope the explanation is understandable

**Comment by melohagan** [comment]
Hi @mattiason 

I don't quite follow. It might help to use real table and field names just so I can wrap my head around the relationships. 

What do your table schemas look like? What do the relationships look like, e.g. one-to-many, many-to-many, etc.

**Comment by mattiason** [comment]
Thanks for you answer and sorry for late reply I was a bit sick

So in one table (A), I have the following

<img width="815" alt="image" src="https://user-images.githubusercontent.com/45860668/164993552-b33a7f87-fac6-4bc9-a4da-5e1cf127e26a.png"> which is the allowed tag types to use.

In the table (B) with tags I have tags 

<img width="1138" alt="image" src="https://user-images.githubusercontent.com/45860668/164993628-fcf94c0f-e1b5-4f34-bd95-0d68f45e945e.png">

Now I would like to create a new table 

In table (C) I would like to add tags from table B but only from the allowed types from table A, but when I select a second tag it should also exclude tags as the option that is of the same type that is already used. So what i want to is to have multi select control or relation from table (b) and only show tags thats is allowed as per table (A) but not exclude those already used in table c.

**Comment by melohagan** [comment]
Hey @mattiason 

Thank you for the additional information. (I hope you're feeling better now!)

I've come up with a solution which hopefully will get you on the right path. My understanding is that you have a setup along the lines of:

- One Task Type -> Many Tasks
- One Task Type -> Many Tags
- Many Tasks -> Many Tags

Then the desire is to make the tags distinct between tasks of the same type. 

In this example, let's assume the 'New Row' Task screen is provided the Task Type/Template through the URL, and the Data Provider filers on this, as seen here: https://docs.budibase.com/docs/url-parameters#navigate-to-with-table-row-buttons

### Step 1 - Data
Add an additional JSON type column to your Task Type table (A) called *Available Tags*
More detail here: https://docs.budibase.com/docs/json

Enter the following schema:
![Screenshot 2022-04-25 at 15 03 49](https://user-images.githubusercontent.com/101575380/165105265-d04bf2f4-70d3-4e95-a770-0debbd1f63f6.png)

![Screenshot 2022-04-25 at 15 04 08](https://user-images.githubusercontent.com/101575380/165105324-a0892a3c-8d15-4089-a0bd-4659ebd126fc.png)

This will keep track of which tags have not already been used for this type, with undefined meaning all are available, and empty array `tags: []` meaning none are available.

### Step 2 - Design
Within the 'New Task' screen, add the following nested structure under the form:
![Screenshot 2022-04-25 at 16 11 27](https://user-images.githubusercontent.com/101575380/165118519-de587b88-9a61-4c61-af45-07bec1467ed9.png)

| Data Provider | Repeater | 
|--------------- | ---------| 
| ![Screenshot 2022-04-25 at 16 13 39](https://user-images.githubusercontent.com/101575380/165118905-b7d98a7d-0b41-4dfd-822d-1cc7c564833c.png) | ![Screenshot 2022-04-25 at 16 13 53](https://user-images.githubusercontent.com/101575380/165118958-73d6b31d-ccce-4b51-9acf-c2f24037c925.png) |
![Screenshot 2022-04-25 at 16 14 32](https://user-images.githubusercontent.com/101575380/165119133-75266b6a-fed5-4d40-83db-fd23134427d1.png) <<Select the tags JSON Array

In the field group, add a multi-select below the relationship picker:
![Screenshot 2022-04-25 at 16 18 37](https://user-images.githubusercontent.com/101575380/165119988-a05de53e-3468-4d95-bf85-4d4af43c1b2c.png)

![Screenshot 2022-04-25 at 16 19 21](https://user-images.githubusercontent.com/101575380/165120147-615c225c-6d33-4a78-951d-12615de4c7f1.png)

**Configure conditions**

| Picker | Condition |
| ------ | --------- | 
| Relationship | ![Screenshot 2022-04-25 at 16 33 24](https://user-images.githubusercontent.com/101575380/165122836-50ff02f9-3760-4225-bc0c-4f3b5ccbc8ca.png) |
| Multi-select | ![Screenshot 2022-04-25 at 16 39 14](https://user-images.githubusercontent.com/101575380/165123901-937ea6cd-2225-46e8-9325-0f5e39ebb991.png) |

### Step 3 - Action
Define actions for the 'Save' button, and add an additional 'Save Row' action. This will update the *Available Tags* column for this task type/template. 

![Screenshot 2022-04-25 at 16 44 34](https://user-images.githubusercontent.com/101575380/165124916-1fa820c6-11cd-4093-afa1-c3a8fc693b50.png)

Under **_id** you could also use the URL binding assuming you've passed that through. 

**JavaScript**
```
let tags = $("Form.Fields.AvailableTags") ?? $("Form.Fields.Tags");
let currentTags = $("Table A Repeater.Task Templates.Available Tags.tags");

//Get all of the allowed tags for this type MINUS any selected types in the form
let notSelectedTags = $("Table A Repeater.Task Templates.Allowed Tags")?.filter(allowed => !tags?.includes(allowed._id));

//Push the tags that were not selected in the picker
if (currentTags?.length > 0) {
	for (let tag in notSelectedTags) {
	  currentTags.push({
			_id: tag._id,
			primaryDisplay: tag.primaryDisplay
		})
	}
} else {
	return notSelectedTags
}

//Remove any existing available tags that were selected in the picker
return currentTags?.filter(tag => !tags.includes(tag._id) && !!tag._id);
```

### Result
![relationship_filter](https://user-images.githubusercontent.com/101575380/165126924-5a9645b1-c145-4577-bfda-59fee014d5d9.gif)

**Reply by mattiason** [reply]
Thank you for a great explanation that is very useful. Still not really solving my use case but I will still try to use it and see if I can think a bit different about it.

**Reply by melohagan** [reply]
If you do find a solution, I would love to know what you did!
Also, if you have any other questions or ideas don't hesitate to post :)
